docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+# SDG API :: Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+block_configs:
+  - block_type: LLMBlock
```

Comment:
It looks like this proposal does not attempt to solve extending the available `block_type` implementations, right? Probably not something that needs to be solved at this point, but it seems that this would be a good opportunity to use a factory pattern of some sort. I can imagine two possible ways to do this:

1. Add a custom `metaclass` to the [Block](https://github.com/instructlab/sdg/blob/main/src/instructlab/sdg/block.py#L17) base class that maintains a class-attribute dict of `name` -> `class` for all available classes, then require users with custom block types to get their blocks imported _before_ instantiating the flow.
2. Allow the `block_type` string to include a qualified import path (e.g. `my_extension.blocks.my_block.MyBlock` and then use `importlib` to do something like `MyBlock = getattr(importlib.import_module(block_type.rsplit(".", 1)[0]), block_type.rsplit(".", 1)[-1])`

Reply:
right - no attempt to allow custom block types right now, just configuring what the lib includes

Reply:
Agree, custom block types is a whole other thing. Definitely would be interested to hear the use case.